src/portable/st/synopsys/dcd_synopsys.c

```diff
@@ -356,17 +357,28 @@ bool dcd_edpt_xfer (uint8_t rhport, uint8_t ep_addr, uint8_t * buffer, uint16_t
   xfer_ctl_t * xfer = XFER_CTL_BASE(epnum, dir);
   xfer->buffer       = buffer;
   xfer->total_len    = total_bytes;
-  xfer->queued_len   = 0;
-  xfer->short_packet = false;
 
   uint16_t num_packets = (total_bytes / xfer->max_size);
-  uint8_t short_packet_size = total_bytes % xfer->max_size;
+  uint8_t const short_packet_size = total_bytes % xfer->max_size;
 
   // Zero-size packet is special case.
   if(short_packet_size > 0 || (total_bytes == 0)) {
     num_packets++;
   }
 
+  // EP0 can only handle one packet
+  if(epnum == 0) {
+    num_packets = 1;
+    if(total_bytes > xfer->max_size) {
+      ep0_pending[dir] = total_bytes - xfer->max_size;
+      total_bytes = xfer->max_size;
+      // Decrement pointer to make EP0 buffers compatible with existing routines.
+      xfer->buffer -= ep0_pending[dir];
```

Comment:
Oh, I wonder how this found it's place in my commit. This was actually the first attempt I made but also thought it'd be not safe. Then instead of decrementing the pointer here I added the incremention of `remaining_bytes`. Doing both is not correct and actually should not work.